### PR TITLE
Implement shifting of UID/GID in ACL xattrs

### LINKED ISSUE #3188: Implement shifting of uid/gid in xattrs
This issue is to track code changes to shared/idmapset to implement the shifting of those uid/gids that are stored in xattrs. This includes:
 - Filesystem ACL entries
 - Unprivileged filesystem capabilities

### PR DESCRIPTION
**Pull request by albertodonato**
Closes #3188

### CODE REVIEW

**Comment by albertodonato** [comment]
I'm not sure how to add a test for this, since the test image doesn't have getfacl/setfacl.
Could we perhaps move to a minimal alpine image for tests?

File: shared/util_linux.go [hunk 1]
```diff
@@ -193,6 +194,69 @@ int shiftowner(char *basepath, char *path, int uid, int gid) {
...
+	if (lstat(path, &sb) > 0) {
+		perror("stat failed");
+		return -1;
+	}
+	if (S_ISLNK(sb.st_mode)) {
+		return 0;
+	}
+
+
+	for (int entry_id = ACL_FIRST_ENTRY; ; entry_id = ACL_NEXT_ENTRY) {
```
Comment (brauner) [inline shared/util_linux.go]: Please declare `int entry_id` above. :)

File: shared/util_linux.go [hunk 2]
```diff
@@ -193,6 +194,69 @@ int shiftowner(char *basepath, char *path, int uid, int gid) {
...
+int shiftacl(char *basepath, char *path, int uid, int gid, int new_uid, int new_gid) {
+	struct stat sb;
+	acl_t acl;
+
+	acl = acl_get_file(path, ACL_TYPE_ACCESS);
+	if (acl == NULL) {
+		return 0;
+	}
+
+	if (lstat(path, &sb) > 0) {
```
Comment (brauner) [inline shared/util_linux.go]: if (lstat() < 0)

File: shared/util_linux.go [hunk 3]
```diff
@@ -36,6 +36,7 @@ import (
 #include <sys/stat.h>
 #include <sys/types.h>
 #include <sys/un.h>
+#include <acl/libacl.h>
```
Comment (brauner) [inline shared/util_linux.go]: `sys/acl.h`

File: shared/util_linux.go [hunk 4]
```diff
@@ -193,6 +194,69 @@ int shiftowner(char *basepath, char *path, int uid, int gid) {
...
+	struct stat sb;
+	acl_t acl;
+
+	acl = acl_get_file(path, ACL_TYPE_ACCESS);
+	if (acl == NULL) {
+		return 0;
+	}
+
+	if (lstat(path, &sb) > 0) {
+		perror("stat failed");
```
Comment (brauner) [inline shared/util_linux.go]: acl_free(acl)

File: shared/util_linux.go [hunk 5]
```diff
@@ -193,6 +194,69 @@ int shiftowner(char *basepath, char *path, int uid, int gid) {
...
+		return -1;
+	}
+	if (S_ISLNK(sb.st_mode)) {
```
Comment (brauner) [inline shared/util_linux.go]: acl_free(acl)

File: shared/util_linux.go [hunk 6]
```diff
@@ -193,6 +194,69 @@ int shiftowner(char *basepath, char *path, int uid, int gid) {
...
+		id_t *id_p;
+		id_t new_id;
+		int update_acl = 0;
+
+		if (acl_get_entry(acl, entry_id, &ent) != 1) {
+			break;
+		}
+
+		if (acl_get_tag_type(ent, &tag) == -1) {
+			perror("failed acl_get_tag_type");
```
Comment (brauner) [inline shared/util_linux.go]: acl_free(acl)

File: shared/util_linux.go (hunk 3 above)
Comment (albertodonato) [inline shared/util_linux.go]: I get undefined references if I use `sys/acl.h`

File: shared/util_linux.go (hunk 2 above)
Comment (albertodonato) [inline shared/util_linux.go]: fixed.

File: shared/util_linux.go (hunk 4 above)
Comment (albertodonato) [inline shared/util_linux.go]: fixed

File: shared/util_linux.go (hunk 5 above)
Comment (albertodonato) [inline shared/util_linux.go]: fixed

File: shared/util_linux.go (hunk 1 above)
Comment (albertodonato) [inline shared/util_linux.go]: why? this is only needd inside the for loop

File: shared/util_linux.go (hunk 6 above)
Comment (albertodonato) [inline shared/util_linux.go]: fixed.

**Comment by albertodonato** [comment]
As mentioned on IRC, jenkins needs libacl1-dev installed for tests to run

**Comment by stgraber** [comment]
Modified the jenkins runner profile to include libacl1-dev, should be rolled out on all of them in an hour or so.

**Comment by albertodonato** [comment]
@stgraber thanks!

**Comment by stgraber** [comment]
jenkins: test this please

File: shared/util_linux.go (hunk 3 above)
Comment (brauner) [inline shared/util_linux.go]: Wait, even with the dev package?

File: shared/util_linux.go [hunk 7]
```diff
@@ -193,6 +194,73 @@ int shiftowner(char *basepath, char *path, int uid, int gid) {
...
+				new_id = *id_p - gid + new_gid;
+				update_acl = 1;
+				break;
+		}
+
+		if (update_acl) {
+			acl_set_qualifier(ent, &new_id);
+			if (acl_set_file(path, ACL_TYPE_ACCESS, acl) == -1) {
+				perror("acl_set_file failed");
+				acl_free(id_p);
```
Comment (brauner) [inline shared/util_linux.go]: You'll likely need to `acl_free(acl)` here too unless `acl_set_file()` claims the memory and frees ith with `acl_free(id_p)`.

File: shared/util_linux.go [hunk 8]
```diff
@@ -193,6 +194,73 @@ int shiftowner(char *basepath, char *path, int uid, int gid) {
...
+				return 1;
+			}
+		}
+		acl_free(id_p);
```
Comment (brauner) [inline shared/util_linux.go]: You'll likely need to `acl_free(acl)` here too unless `acl_set_file()` claims the memory and frees ith with `acl_free(id_p)`.

File: shared/util_linux.go [hunk 9]
```diff
@@ -193,6 +194,73 @@ int shiftowner(char *basepath, char *path, int uid, int gid) {
...
+	}
+	return 0;
```
Comment (brauner) [inline shared/util_linux.go]: You'll likely need to `acl_free(acl)` here too unless `acl_set_file()` claims the memory and frees ith with `acl_free(id_p)`.

**Comment by stgraber** [comment]
So this approach is going to be a problem with non-contiguous uid/gid maps won't it?

File: shared/util_linux.go (hunk 3 above)
Comment (albertodonato) [inline shared/util_linux.go]: Yes. FWIW both the sample code I found on the internet and get/setfacl include both acl.h and libacl.h. What's the issue with including libacl.h?

File: shared/util_linux.go (hunk 7 above)
Comment (albertodonato) [inline shared/util_linux.go]: argh, I actually meant to `acl_free(acl)`, both here and below

**Comment by albertodonato** [comment]
@stgraber @brauner mind having another look?
I rewrote the logic in Go since I needed access to the uid/gid mapping function. IIUC this 
For some reason, now just including `sys/acl.h` works.

File: shared/util_linux.go [hunk 10]
```diff
@@ -252,6 +253,65 @@ func ShiftOwner(basepath string, path string, uid int, gid int) error {
...
+		return nil
+	}
+	defer C.acl_free(unsafe.Pointer(acl))
+
+	for entryId := C.ACL_FIRST_ENTRY; ; entryId = C.ACL_NEXT_ENTRY {
+		var ent C.acl_entry_t
+		var tag C.acl_tag_t
+		updateACL := false
+		// 	id_t *id_p;
+		// 	id_t new_id;
```
Comment (brauner) [inline shared/util_linux.go]: Can these be removed? :)

File: shared/util_linux.go [hunk 11]
```diff
@@ -252,6 +253,65 @@ func ShiftOwner(basepath string, path string, uid int, gid int) error {
...
+
+		if C.acl_get_entry(acl, C.int(entryId), &ent) != 1 {
```
Comment (brauner) [inline shared/util_linux.go]: Can you make this:

```
ret = C.fun()
ret != 1
```

and similar for `C.acl_get_tag_type()`, please? :)

File: shared/util_linux.go [hunk 12]
```diff
@@ -252,6 +253,65 @@ func ShiftOwner(basepath string, path string, uid int, gid int) error {
...
+			newId, _ = shiftIds((int64)(*idp), -1)
+			updateACL = true
+
+		case C.ACL_GROUP:
+			_, newId = shiftIds(-1, (int64)(*idp))
+			updateACL = true
+		}
+
+		if updateACL {
+			C.acl_set_qualifier(ent, unsafe.Pointer(&newId))
```
Comment (brauner) [inline shared/util_linux.go]: Why are you ignoring the return value here?

File: shared/util_linux.go (hunk 11 above)
Comment (brauner) [inline shared/util_linux.go]: and also for `C.acl_set_file()` :)

File: shared/util_linux.go (hunk 12 above)
Comment (albertodonato) [inline shared/util_linux.go]: Added

File: shared/util_linux.go (hunk 10 above)
Comment (albertodonato) [inline shared/util_linux.go]: removed

File: shared/util_linux.go (hunk 11 above)
Comment (albertodonato) [inline shared/util_linux.go]: Done, although I find it way nicer on a single line

**Comment by albertodonato** [comment]
jenkins: test this please

**Comment by albertodonato** [comment]
jenkins: test this please

**Pull request merged**